### ISSUE TITLE
Editor: restore Exclude Room feature

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#2119]
**Problem**

I was asked about this by somebody (possibly by Dave Gilbert) back during 3.6.0 development, but got distracted and never got to do this or write a ticket at least.

Back in AGS 3.2.1 there have been a context menu command for Rooms called "Exclude from game" which allowed to remove particular room(s) from compilation. This was useful to make partial game builds, like demos.

Unfortunately, it was implemented in a silly way: instead of tagging a Room item as "excluded" it just removed a room from project, without deleting the room file. User would have to import room back when necessary. This was probably a reason why this command was mistakenly thought to be unnecessary and removed at some point.

The code may still be found in the very early commits in this repository, but given it never worked right, I recommend to ignore them, and instead implement this correctly anew.

**Proposal**

First of all, in my opinion the "excluded" state should not be a part of the Room class nor game data overall. I see this rather as a part of a compilation config. Therefore, I think it should be a part of the "workspace" state, the one that is saved in Game.agf.user.

The tasks are following:
1. Implement room excluded states, as part of the workspace settings. This may be done either as some class similar to [RoomDesignData](https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Editor/Components/RoomDesignData.cs),  or as a list of excluded rooms in a [WorkspaceState](https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Types/WorkspaceState.cs), or else. This data will be saved in Game.agf.user.
2. Add "Exclude from game" and "Include to game" context menu commands for Rooms in the Project Explorer.
3. Implement some way to visually indicate an excluded room. For example: make it appear greyed out.
4. Skip the excluded rooms during game compilation, regardless of the build type (test run or full rebuild).

**Comment by ericoporto** [comment]
I think the best approach here is to create some tagging where the default includes all rooms but you can also just tag a few rooms as "demo game" and then somewhere set to include only rooms with tag "demo game" for building.

Possibly it could also have the exclude rooms by tag so when you are doing your "final" build you could exclude your debug rooms, but I think this is less common as games usually ship with most of their debugging so devs can figure why some issue an user has happens.

**Comment by ivan-mogilko** [comment]
Right, that sounds much better, as this way user won't have to do this by hand all the time.

In such way, however, I think there has to be some visual cue for tagged items, which may be toggled on and off. For example, in the project tree's context menu, there could be a submenu called "display tags" (or else) with checkable items inside, and rooms will be shown greyed out or otherwise marked if they do or do not have checked tags.

Alternatively, this menu may work as an intermediate (temporary) solution for selecting tags for building.

**Comment by ericoporto** [comment]
I have been playing with the idea of a rooms overview panel, that would allow to browse the rooms graphically, with a thumbnail per each room background with some overlayed information. I saw this in a sci Editor for Sierra games and thought it made much easier to browse when you have lots of rooms. I was trying to do it in ags3 but I think I will rework for ags4 because the backgrounds being their own file makes the code much simpler.

**Comment by ivan-mogilko** [comment]
> I have been playing with the idea of a rooms overview panel, that would allow to browse the rooms graphically, with a thumbnail per each room background with some overlayed information.

Hmm, not criticizing the idea of such panel itself, but I don't really think this should be a solution for this exact problem. We have a Project Explorer that serves an overview of the game contents, and my belief is that it should reflect the game building. This is why I don't think it's a good idea to rely solely on an additional panel for this.

Besides, if "tag for build" feature is implemented for rooms, it's possible someone would like to implement this for some other things too.

The room "tagging" idea seem simple enough to be added or ported to ags3 as well. As much as I can see, the minimal implementation would require following:
1. Have "room tag" properties serialized in either game or workspace settings. Previously I meant "room exclusiion" as a part of workspace, but with the "tags" it seems more suitable as a part of the game project itself.
2. In a room item context menu have commands:
    - edit list of tags (opens a simple window with a list and add/remove/remove all buttons);
    - assign tag: opens a submenu with checkable tag items.
3. Somewhere else, either in a context menu for "Rooms" parent node, or Build menu, have also a submenu with checkable list of tags to include in a build, as well as "select all".
4. When compiling, the Editor selects only rooms which match the tags.

Does this makes sense?

**Comment by ericoporto** [comment]
Yes it makes sense. About serializing in the AGS project itself, it could be in the Unloaded rooms section in Game.agf - I don't know how the Editor actually writes things there, but it's from where I pulled the room description recently (in the description in tab PR).

**Comment by ericoporto** [comment]
Just two small comments (I am asking questions because in general the scope looks like something I can implement so I am just discussing the design before attempting something)

Microsoft Teams has tags for tasks in a task board, and it works like this: there are 16 pre-existing tags of different colors, with names matching the colors. You can, per task board, rename a tag but you can't modify the colors and neither add or remove tags. Each task can have any number of tags. I guess this was done as this because it was quicker to implement - and it's been like this for three years.

The second thing is I was thinking if it's better to have the information of room-taglist be stored as each tag containing a list of room numbers or each room containing a list of tag IDs.

**Comment by ericoporto** [comment]
Found a simple code example  for tags like drawing (ignore the usage of the `Tag` property from TreeNode which is a different thing)

https://learn.microsoft.com/en-us/dotnet/api/system.windows.forms.treeview.drawnode?view=netframework-4.6.2&redirectedfrom=MSDN

![image](https://github.com/adventuregamestudio/ags/assets/2244442/4d5cb2e7-36d0-4ebb-b1f5-4a3b9f6d4cfd)

But it does so by overriding the DrawNode in the TreeView. I think this is done because the area of the node itself has the exact size of the text+icon (wouldn't fit additional info).
